In [30]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

In [2]:
df = pd.read_csv("hack_dataset.csv")

In [3]:
df.head()

surname  creditscore  age  tenure    balance  numofproducts  hascrcard  \
0     2023          668   33       3       0.00              2          1   
1     2024          627   33       1       0.00              2          1   
2     1236          678   40      10       0.00              2          1   
3     1362          581   34       2  148882.54              1          1   
4      491          716   33       5       0.00              2          1   

   isactivemember  estimatedsalary  exited  ...  france  germany  spain  \
0               0        181449.97       0  ...     1.0      0.0    0.0   
1               1         49503.50       0  ...     1.0      0.0    0.0   
2               0        184866.69       0  ...     1.0      0.0    0.0   
3               1         84560.88       0  ...     1.0      0.0    0.0   
4               1         15068.83       0  ...     0.0      0.0    1.0   

   female  male  mem__no__products  cred_bal_sal  bal_sal  tenure_age  \
0     0.0   1.0                0.0          0.00     0.00        0.09   
1     0.0   1.0                2.0          0.00     0.00        0.03   
2     0.0   1.0                0.0          0.00     0.00        0.25   
3     0.0   1.0                1.0       1022.94     1.76        0.06   
4     0.0   1.0                2.0          0.00     0.00        0.15   

   age_tenure_product  
0                99.0  
1                33.0  
2               400.0  
3                68.0  
4               165.0  

[5 rows x 25 columns]

In [4]:
df.tail()

surname  creditscore  age  tenure    balance  numofproducts  hascrcard  \
465      482          532   24       8  142755.25              1          0   
466     1943          714   29       1       0.00              2          1   
467     1609          686   34       2       0.00              2          1   
468     2089          543   35       7       0.00              2          0   
469     2662          641   55       8       0.00              2          1   

     isactivemember  estimatedsalary  exited  ...  france  germany  spain  \
465               1         34231.48       0  ...     0.0      0.0    1.0   
466               1         86479.49       0  ...     1.0      0.0    0.0   
467               0        127569.80       0  ...     0.0      0.0    1.0   
468               1         56767.91       0  ...     1.0      0.0    0.0   
469               0         91285.88       1  ...     NaN      NaN    NaN   

     female  male  mem__no__products  cred_bal_sal  bal_sal  tenure_age  \
465     0.0   1.0                1.0        2218.6     4.17        0.33   
466     0.0   1.0                2.0           0.0     0.00        0.03   
467     0.0   1.0                0.0           0.0     0.00        0.06   
468     0.0   1.0                2.0           0.0     0.00        0.20   
469     NaN   NaN                NaN           NaN      NaN         NaN   

     age_tenure_product  
465               192.0  
466                29.0  
467                68.0  
468               245.0  
469                 NaN  

[5 rows x 25 columns]

In [6]:
df.describe()

surname  creditscore         age      tenure        balance  \
count   470.000000   470.000000  470.000000  470.000000     470.000000   
mean   1683.157447   654.114894   37.848936    5.102128   58353.392702   
std     791.281860    83.497785    8.787988    3.001456   63650.116656   
min      27.000000   359.000000   20.000000    0.000000       0.000000   
25%    1081.000000   594.000000   32.000000    2.000000       0.000000   
50%    1746.000000   663.000000   37.000000    5.000000       0.000000   
75%    2408.500000   708.750000   42.000000    8.000000  123021.805000   
max    2931.000000   850.000000   71.000000   10.000000  190566.650000   

       numofproducts   hascrcard  isactivemember  estimatedsalary      exited  \
count     470.000000  470.000000      470.000000       470.000000  470.000000   
mean        1.574468    0.753191        0.529787    114573.016277    0.200000   
std         0.574686    0.431614        0.499644     48736.408435    0.400426   
min         1.000000    0.000000        0.000000       933.380000    0.000000   
25%         1.000000    1.000000        0.000000     79583.960000    0.000000   
50%         2.000000    1.000000        1.000000    120976.240000    0.000000   
75%         2.000000    1.000000        1.000000    152816.120000    0.000000   
max         4.000000    1.000000        1.000000    199929.170000    1.000000   

       ...      france     germany       spain      female        male  \
count  ...  469.000000  469.000000  469.000000  469.000000  469.000000   
mean   ...    0.614072    0.181237    0.204691    0.413646    0.586354   
std    ...    0.487333    0.385625    0.403906    0.493012    0.493012   
min    ...    0.000000    0.000000    0.000000    0.000000    0.000000   
25%    ...    0.000000    0.000000    0.000000    0.000000    0.000000   
50%    ...    1.000000    0.000000    0.000000    0.000000    1.000000   
75%    ...    1.000000    0.000000    0.000000    1.000000    1.000000   
max    ...    1.000000    1.000000    1.000000    1.000000    1.000000   

       mem__no__products  cred_bal_sal     bal_sal  tenure_age  \
count         469.000000    469.000000  469.000000  469.000000   
mean            0.835821    713.949467    1.081109    0.141919   
std             0.887181   3050.974373    4.484300    0.089130   
min             0.000000      0.000000    0.000000    0.000000   
25%             0.000000      0.000000    0.000000    0.070000   
50%             1.000000      0.000000    0.000000    0.140000   
75%             2.000000    724.680000    1.070000    0.210000   
max             4.000000  40509.020000   59.750000    0.400000   

       age_tenure_product  
count          469.000000  
mean           192.085288  
std            125.045997  
min              0.000000  
25%             84.000000  
50%            175.000000  
75%            288.000000  
max            639.000000  

[8 rows x 25 columns]

In [8]:
df.shape

(470, 25)

In [38]:
df.columns

Index(['creditscore', 'age', 'tenure', 'balance', 'numofproducts', 'hascrcard',
       'isactivemember', 'estimatedsalary', 'exited', 'surname_tfidf_0',
       'surname_tfidf_1', 'surname_tfidf_2', 'surname_tfidf_3',
       'surname_tfidf_4', 'france', 'germany', 'spain', 'female', 'male',
       'mem__no__products', 'cred_bal_sal', 'bal_sal', 'tenure_age',
       'age_tenure_product', 'gender_label', 'geography_label'],
      dtype='object')

In [17]:
y = df['exited']

In [18]:
attributes = ['creditscore', 'age', 'tenure', 'balance', 'numofproducts', 'hascrcard',
    'isactivemember', 'estimatedsalary', 'surname_tfidf_0', 'surname_tfidf_1',
    'surname_tfidf_2', 'surname_tfidf_3', 'surname_tfidf_4',
    'mem__no__products', 'cred_bal_sal', 'bal_sal', 'tenure_age', 'age_tenure_product',
    'gender_label', 'geography_label']

In [19]:
X = df[[col for col in attributes if col in df.columns]]

In [34]:
categorical = ['gender_label', 'geography_label', 'numofproducts']
categorical = [col for col in categorical_cols if col in X.columns]

In [35]:
X = pd.get_dummies(X, columns=categorical, drop_first=True)

In [36]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [37]:
model = LogisticRegression(solver='liblinear', random_state=42, max_iter=1000)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000, random_state=42, solver='liblinear')

In [27]:
y_pred = model.predict(X_test)

In [31]:
print(classification_report(y_test, y_pred, target_names=['Stayed (0)', 'Exited (1)']))

              precision    recall  f1-score   support

  Stayed (0)       0.82      0.99      0.90        75
  Exited (1)       0.75      0.16      0.26        19

    accuracy                           0.82        94
   macro avg       0.79      0.57      0.58        94
weighted avg       0.81      0.82      0.77        94



In [33]:
accuracy = accuracy_score(y_test, y_pred)
accuracy

0.8191489361702128

This Machine Learning model is a Logistic Regression Classifier designed to predict customer churn. It achieved an overall accuracy of 82%, meaning it correctly identified whether a customer would stay or leave most of the time. While it's highly effective at predicting customers who will stay.